src/include/hardware.h

```diff
@@ -31,19 +31,40 @@
  * same memory location, but to do this safely we need to precisely control how reads to memory are
  * performed. This macro gives us control over this.
  *
- * GCC and Clang have a __typeof__ compiler builtin which allows us to temporarily cast the value
- * being read as a volatile and achieve volatile semantics. For other compilers we'll fall back on
- * inserting a read barrier after the read (our pre-existing implementation) which prevents invented
- * and fused loads for this variable in the code following the expression.
+ * To allow this macro to be used in a backwards compatible way it has been implemented using
+ * barriers. Technically this can be achieved by a volatile cast.
+ */
+#define WT_READ_ONCE(v, val) \
+    do {                     \
+        WT_READ_BARRIER();   \
+        (v) = (val);         \
+        WT_READ_BARRIER();   \
+    } while (0)
+
+/*
+ * WT_WRITE_ONCE --
+ *
+ * Ensure a single write to memory in the source code produces a single write to memory in the
+ * compiled output.
  *
- * FIXME-WT-11718 - Once Windows build machines that support C11 _Generics are available this macro
- * will be updated to use _Generic on all platforms.
+ * We can describe a scenario where a variable in memory is read to a local register, and then the
+ * register value is repeatedly updated before the final value is written back to memory. Without a
+ * WT_WRITE_ONCE wrapping the final write back to memory, the compiler is allowed to convert these
+ * register writes into writes to memory (see "register spilling") which can leak interim, invalid,
+ * state to other threads. We call these 'invented stores' in WiredTiger.
+ *
+ * See the read once macro description for more details.
+ *
+ * To allow this macro to be used in a backwards compatible way it has been implemented using
+ * barriers. Technically this can be achieved by a volatile cast.
  */
-#if defined(__GNUC__) || defined(__clang__)
-#define WT_READ_ONCE(v, val) (v) = (*(volatile __typeof__(val) *)&(val))
-#else
-#define WT_READ_ONCE(v, val) WT_ORDERED_READ(v, val)
-#endif
+#define WT_WRITE_ONCE(v, val) \
+    do {                      \
+        WT_WRITE_BARRIER();   \
```

Comment:
This does surprise me - adding a write barrier to the final write stopping prior "invented stores" isn't intuitive - though a lot of this isn't intuitive.
The bi-directional barriers feel like overkill - could you help me understand when they are necessary?

Reply:
I'll reply to this comment in the above comment so everything is in one place.